test/library/library.go

```
@@ -32,3 +47,117 @@ func GenerateNameForTest(t *testing.T, prefix string) string {
+// The number of pods should stabilize in that period of time.
+func areAPIServersOnTheSameRevision(podClient corev1client.PodInterface) (bool, error) {
+	revisionLabel := "revision"
+
+	// do a live list so we never get confused about what revision we are on
+	apiServerPods, err := podClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "apiserver=true"})
+	if err != nil {
+		return false, err
+	}
+
+	good, bad, progressing, err := categorizePods(apiServerPods.Items)
+	if err != nil {
+		return false, err
+	}
+	if progressing {
+		return false, nil
+	}
+
+	goodRevisions := revisions(revisionLabel, good)
+	failingRevisions := revisions(revisionLabel, bad)
+
+	if len(goodRevisions) != 1 {
+		return false, nil // api servers have not converged onto a single revision
+	}
+	revision, _ := goodRevisions.PopAny()
+	if len(revision) == 0 {
+		revision = "0"
+	}
+
+	if failingRevisions.Has(revision) {
+		return false, fmt.Errorf("api server revision %s has both running and failed pods", revision)
+	}
+
+	return true, nil
+}
+
+func revisions(revisionLabel string, pods []*corev1.Pod) sets.String {
+	ret := sets.NewString()
+	for _, p := range pods {
+		ret.Insert(p.Labels[revisionLabel])
+	}
+	return ret
+}
+
+func categorizePods(pods []corev1.Pod) (good []*corev1.Pod, bad []*corev1.Pod, progressing bool, err error) {
+	if len(pods) == 0 {
+		return nil, nil, true, err
+	}
+	for _, apiServerPod := range pods {
+		switch phase := apiServerPod.Status.Phase; phase {
+		case corev1.PodRunning:
+			if !podReady(apiServerPod) {
+				return nil, nil, true, nil // pods are not fully ready
+			}
+			goodPod := apiServerPod // shallow copy because apiServerPod is bound loop var
+			good = append(good, &goodPod)
+		case corev1.PodPending:
+			return nil, nil, true, nil // pods are not fully ready
+		case corev1.PodUnknown:
+			return nil, nil, false, fmt.Errorf("api server pod %s in unknown phase", apiServerPod.Name)
```

Is this condition unrecoverable?

---

I don't think so. In a case like that, I would rather return to fail early and let a human fix the issue - possibly by adding a new condition.